[3006.x] Make deb scripts more compliant with Debian policy

baby-gnu opened the pull request:
### What does this PR do?

- make the deb scripts manage systemd status during upgrade
- improve script compliance with debian policy based on [dh-make examples](https://salsa.debian.org/debian/dh-make)

### What issues does this PR fix or reference?

Fixes #66507

### Previous Behavior

When upgrading debian salt packages, the current status of the services are not managed:

- previously disabled packages became enable
- masked packages make the upgrade fail

<details><summary>Status of the services before upgrade</summary>

```
○ salt-master.service - The Salt Master Server
     Loaded: loaded (/lib/systemd/system/salt-master.service; disabled; preset: enabled)
     Active: inactive (dead)
       Docs: man:salt-master(1)
             file:///usr/share/doc/salt/html/contents.html
             https://docs.saltproject.io/en/latest/contents.html

○ salt-minion.service - The Salt Minion
     Loaded: loaded (/lib/systemd/system/salt-minion.service; disabled; preset: enabled)
     Active: inactive (dead)
       Docs: man:salt-minion(1)
             file:///usr/share/doc/salt/html/contents.html
             https://docs.saltproject.io/en/latest/contents.html

○ salt-api.service
     Loaded: masked (Reason: Unit salt-api.service is masked.)
     Active: inactive (dead)
```
</details>

<details><summary>Error log during upgrade</summary>

```
```
</details>

<details><summary>Status of the services after failed upgrade</summary>

```
○ salt-master.service - The Salt Master Server
     Loaded: loaded (/lib/systemd/system/salt-master.service; enabled; preset: enabled)
     Active: inactive (dead)
       Docs: man:salt-master(1)
             file:///usr/share/doc/salt/html/contents.html
             https://docs.saltproject.io/en/latest/contents.html

○ salt-minion.service - The Salt Minion
     Loaded: loaded (/lib/systemd/system/salt-minion.service; disabled; preset: enabled)
     Active: inactive (dead)
       Docs: man:salt-minion(1)
             file:///usr/share/doc/salt/html/contents.html
             https://docs.saltproject.io/en/latest/contents.html

○ salt-api.service
     Loaded: masked (Reason: Unit salt-api.service is masked.)
     Active: inactive (dead)
```
</details>

### New Behavior

The upgrade should work with masked services and keep previous status after upgrade.

I can't test it because I did not find the way to build custom packages to test

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices, including the
[PR Guidelines](https://docs.saltproject.io/en/master/topics/development/pull_requests.html).

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

baby-gnu commented:
To me, my PR is working fire, the generated scripts now looks like this:


<details><summary><code>salt-master.preinst</code></summary>

``` bash
#!/bin/sh
# preinst script for salt-package.
#
# See: dh_installdeb(1).

set -e

# Summary of how this script can be called:
#        * <new-preinst> 'install'
#        * <new-preinst> 'install' <old-version>
#        * <new-preinst> 'upgrade' <old-version>
#        * <old-preinst> 'abort-upgrade' <new-version>
# for details, see https://www.debian.org/doc/debian-policy/ or
# the debian-policy package.


case "$1" in
  install|upgrade)
    [ -z "$SALT_HOME" ] && SALT_HOME=/opt/saltstack/salt
    [ -z "$SALT_USER" ] && SALT_USER=salt
    [ -z "$SALT_NAME" ] && SALT_NAME="Salt"
    [ -z "$SALT_GROUP" ] && SALT_GROUP=salt
    PY_VER=$(/opt/saltstack/salt/bin/python3 -c "import sys; sys.stdout.write('{}.{}'.format(*sys.version_info)); sys.stdout.flush();")

    # Reset permissions to fix previous installs
    find ${SALT_HOME} /etc/salt /var/log/salt /var/cache/salt /var/run/salt \
        \! \( -path /etc/salt/cloud.deploy.d\* -o -path /var/log/salt/cloud -o -path /opt/saltstack/salt/lib/python${PY_VER}/site-packages/salt/cloud/deploy\* \) -a \
        \( -user ${SALT_USER} -o -group ${SALT_GROUP} \) -exec chown root:root \{\} \;

  # remove incorrectly installed ufw salt-master directory - issue 57712
  test -d /etc/ufw/applications.d/salt-master && rm -rf /etc/ufw/applications.d/salt-master || /bin/true

  ;;

  abort-upgrade)
  ;;

  *)
    echo "preinst called with unknown argument '$1'" >&2
    exit 1
  ;;
esac

# dh_installdeb will replace this with shell code automatically
# generated by other debhelper scripts.



exit 0
```
</details>

<details><summary><code>salt-master.postinst</code></summary>

``` bash
#!/bin/sh
# postinst script for salt-master
#
# see: dh_installdeb(1)

set -e

# summary of how this script can be called:
#        * <postinst> `configure' <most-recently-configured-version>
#        * <old-postinst> `abort-upgrade' <new version>
#        * <conflictor's-postinst> `abort-remove' `in-favour' <package>
#          <new-version>
#        * <postinst> `abort-remove'
#        * <deconfigured's-postinst> `abort-deconfigure' `in-favour'
#          <failed-install-package> <version> `removing'
#          <conflicting-package> <version>
# for details, see https://www.debian.org/doc/debian-policy/ or
# the debian-policy package

case "$1" in
  configure)
    if [ ! -e "/var/log/salt/master" ]; then
      touch /var/log/salt/master
      chmod 640 /var/log/salt/master
    fi
    if [ ! -e "/var/log/salt/key" ]; then
      touch /var/log/salt/key
      chmod 640 /var/log/salt/key
    fi
    chown -R salt:salt /etc/salt/pki/master /etc/salt/master.d /var/log/salt/master /var/log/salt/key /var/cache/salt/master /var/run/salt/master
  ;;
  abort-upgrade|abort-remove|abort-deconfigure)
  ;;

  *)
    echo "postinst called with unknown argument '$1'" >&2
    exit 1
  ;;
esac

# dh_installdeb will replace this with shell code automatically
# generated by other debhelper scripts.

# Automatically added by dh_systemd_enable/13.11.4
if [ "$1" = "configure" ] || [ "$1" = "abort-upgrade" ] || [ "$1" = "abort-deconfigure" ] || [ "$1" = "abort-remove" ] ; then
	# The following line should be removed in trixie or trixie+1
	deb-systemd-helper unmask 'salt-master.service' >/dev/null || true

	# was-enabled defaults to true, so new installations run enable.
	if deb-systemd-helper --quiet was-enabled 'salt-master.service'; then
		# Enables the unit on first installation, creates new
		# symlinks on upgrades if the unit file has changed.
		deb-systemd-helper enable 'salt-master.service' >/dev/null || true
	else
		# Update the statefile to add new symlinks (if any), which need to be
		# cleaned up on purge. Also remove old symlinks.
		deb-systemd-helper update-state 'salt-master.service' >/dev/null || true
	fi
fi
# End automatically added section
# Automatically added by dh_systemd_start/13.11.4
if [ "$1" = "configure" ] || [ "$1" = "abort-upgrade" ] || [ "$1" = "abort-deconfigure" ] || [ "$1" = "abort-remove" ] ; then
	if [ -d /run/systemd/system ]; then
		systemctl --system daemon-reload >/dev/null || true
		if [ -n "$2" ]; then
			_dh_action=restart
		else
			_dh_action=start
		fi
		deb-systemd-invoke $_dh_action 'salt-master.service' >/dev/null || true
	fi
fi
# End automatically added section


exit 0
```
</details>

<details><summary><code>salt-master.prerm</code></summary>

``` bash
#!/bin/sh
set -e
# Automatically added by dh_systemd_start/13.11.4
if [ -z "${DPKG_ROOT:-}" ] && [ "$1" = remove ] && [ -d /run/systemd/system ] ; then
	deb-systemd-invoke stop 'salt-master.service' >/dev/null || true
fi
# End automatically added section

```
</details>

<details><summary><code>salt-master.postrm</code></summary>

``` bash
#!/bin/sh
set -e
# Automatically added by dh_systemd_start/13.11.4
if [ "$1" = remove ] && [ -d /run/systemd/system ] ; then
	systemctl --system daemon-reload >/dev/null || true
fi
# End automatically added section
# Automatically added by dh_systemd_enable/13.11.4
if [ "$1" = "purge" ]; then
	if [ -x "/usr/bin/deb-systemd-helper" ]; then
		deb-systemd-helper purge 'salt-master.service' >/dev/null || true
	fi
fi
# End automatically added section
```
</details>

baby-gnu commented:
I fixed the `find` in `salt-master.preinst` which do try to run on nonexistent directory if service are disabled or masked.

baby-gnu commented:
My latest test is working fine:

<details><summary><code>systemctl status salt-master.service salt-minion.service salt-api.service salt-syndic.service</code></summary>

```
○ salt-master.service
     Loaded: masked (Reason: Unit salt-master.service is masked.)
     Active: inactive (dead)

○ salt-minion.service
     Loaded: masked (Reason: Unit salt-minion.service is masked.)
     Active: inactive (dead)

○ salt-api.service
     Loaded: masked (Reason: Unit salt-api.service is masked.)
     Active: inactive (dead)

○ salt-syndic.service
     Loaded: masked (Reason: Unit salt-syndic.service is masked.)
     Active: inactive (dead)
```
</details>

<details><summary><code>apt install ./salt*.deb</code></summary>

```
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'salt-api' instead of './salt-api_3007.1+211.g477aae15c6_amd64.deb'
Note, selecting 'salt-cloud' instead of './salt-cloud_3007.1+211.g477aae15c6_amd64.deb'
Note, selecting 'salt-common' instead of './salt-common_3007.1+211.g477aae15c6_amd64.deb'
Note, selecting 'salt-master' instead of './salt-master_3007.1+211.g477aae15c6_amd64.deb'
Note, selecting 'salt-minion' instead of './salt-minion_3007.1+211.g477aae15c6_amd64.deb'
Note, selecting 'salt-ssh' instead of './salt-ssh_3007.1+211.g477aae15c6_amd64.deb'
Note, selecting 'salt-syndic' instead of './salt-syndic_3007.1+211.g477aae15c6_amd64.deb'
The following packages will be upgraded:
  salt-api salt-cloud salt-common salt-master salt-minion salt-ssh salt-syndic
7 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/34.3 MB of archives.
After this operation, 16.3 MB disk space will be freed.
Get:1 /root/salt/salt-syndic_3007.1+211.g477aae15c6_amd64.deb salt-syndic amd64 3007.1+211.g477aae15c6 [78.6 kB]
Get:2 /root/salt/salt-cloud_3007.1+211.g477aae15c6_amd64.deb salt-cloud amd64 3007.1+211.g477aae15c6 [81.3 kB]
Get:3 /root/salt/salt-ssh_3007.1+211.g477aae15c6_amd64.deb salt-ssh amd64 3007.1+211.g477aae15c6 [80.4 kB]
Get:4 /root/salt/salt-master_3007.1+211.g477aae15c6_amd64.deb salt-master amd64 3007.1+211.g477aae15c6 [107 kB]
Get:5 /root/salt/salt-common_3007.1+211.g477aae15c6_amd64.deb salt-common amd64 3007.1+211.g477aae15c6 [33.8 MB]
Get:6 /root/salt/salt-minion_3007.1+211.g477aae15c6_amd64.deb salt-minion amd64 3007.1+211.g477aae15c6 [94.0 kB]
Get:7 /root/salt/salt-api_3007.1+211.g477aae15c6_amd64.deb salt-api amd64 3007.1+211.g477aae15c6 [78.8 kB]
(Reading database ... 197699 files and directories currently installed.)
Preparing to unpack .../0-salt-syndic_3007.1+211.g477aae15c6_amd64.deb ...
Unpacking salt-syndic (3007.1+211.g477aae15c6) over (3007.1) ...
Preparing to unpack .../1-salt-cloud_3007.1+211.g477aae15c6_amd64.deb ...
Unpacking salt-cloud (3007.1+211.g477aae15c6) over (3007.1) ...
Preparing to unpack .../2-salt-ssh_3007.1+211.g477aae15c6_amd64.deb ...
Unpacking salt-ssh (3007.1+211.g477aae15c6) over (3007.1) ...
Preparing to unpack .../3-salt-master_3007.1+211.g477aae15c6_amd64.deb ...
Unpacking salt-master (3007.1+211.g477aae15c6) over (3007.1) ...
Preparing to unpack .../4-salt-common_3007.1+211.g477aae15c6_amd64.deb ...
usermod: no changes
Unpacking salt-common (3007.1+211.g477aae15c6) over (3007.1) ...
Preparing to unpack .../5-salt-minion_3007.1+211.g477aae15c6_amd64.deb ...
Unpacking salt-minion (3007.1+211.g477aae15c6) over (3007.1) ...
Preparing to unpack .../6-salt-api_3007.1+211.g477aae15c6_amd64.deb ...
Unpacking salt-api (3007.1+211.g477aae15c6) over (3007.1) ...
Setting up salt-common (3007.1+211.g477aae15c6) ...
Setting up salt-ssh (3007.1+211.g477aae15c6) ...
Setting up salt-cloud (3007.1+211.g477aae15c6) ...
Setting up salt-minion (3007.1+211.g477aae15c6) ...
salt-minion.service is a disabled or a static unit not running, not starting it.
Setting up salt-master (3007.1+211.g477aae15c6) ...
Failed to preset unit, unit /etc/systemd/system/salt-master.service is masked.
/usr/bin/deb-systemd-helper: error: systemctl preset failed on salt-master.service: No such file or directory
salt-master.service is a disabled or a static unit not running, not starting it.
Setting up salt-syndic (3007.1+211.g477aae15c6) ...
salt-syndic.service is a disabled or a static unit not running, not starting it.
Setting up salt-api (3007.1+211.g477aae15c6) ...
Failed to preset unit, unit /etc/systemd/system/salt-api.service is masked.
/usr/bin/deb-systemd-helper: error: systemctl preset failed on salt-api.service: No such file or directory
salt-api.service is a disabled or a static unit not running, not starting it.
Processing triggers for man-db (2.12.0-4build2) ...
Processing triggers for libc-bin (2.39-0ubuntu8.2) ...
Scanning processes...                                                                                                                                                                    
Scanning linux images...                                                                                                                                                                 

Running kernel seems to be up-to-date.

No services need to be restarted.

No containers need to be restarted.

No user sessions are running outdated binaries.

No VM guests are running outdated hypervisor (qemu) binaries on this host.
N: Download is performed unsandboxed as root as file '/root/salt/salt-syndic_3007.1+211.g477aae15c6_amd64.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)
```
</details>

<details><summary><code>systemctl status salt-master.service salt-minion.service salt-api.service salt-syndic.service</code></summary>

```
○ salt-master.service
     Loaded: masked (Reason: Unit salt-master.service is masked.)
     Active: inactive (dead)

○ salt-minion.service
     Loaded: masked (Reason: Unit salt-minion.service is masked.)
     Active: inactive (dead)

○ salt-api.service
     Loaded: masked (Reason: Unit salt-api.service is masked.)
     Active: inactive (dead)

○ salt-syndic.service
     Loaded: masked (Reason: Unit salt-syndic.service is masked.)
     Active: inactive (dead)
```
</details>

dmurphy18 reviewed (dismissed):
Will need to watch the interaction of these changes in master branch when merge forwarding PR 66218 from 3006.x branch. If there is a better way to handle the systemd state, all ears

baby-gnu commented:
> Will need to watch the interaction of these changes in master branch when merge forwarding PR 66218 from 3006.x branch. If there is a better way to handle the systemd state, all ears

My patch only deal with Debian systems by using their standard way of dealing with this issue, I don't know if other distributions have their way. I would prefer my PR which will evolve with de debian tooling instead of a custom solution, if my opinion can count :wink:

akshar-rafay commented:
Hi team,

Any update on this? Our buildx multiarch docker builds for salt-master are failing with the following error 
```
Error while loading /var/lib/dpkg/tmp.ci/preinst: Exec format error
dpkg: error processing archive /var/cache/apt/archives/salt-master_3006.8_arm64.deb (--unpack)
```
Our Dockerfile is running ```apt-get install salt-common salt-master salt-api```
This is due to the missing shebang (```#!/bin/sh```) and ```set -e```. 
Manually updating the Dockerfile to download the debian, extract it, add the shebang and ```set -e``` to the preinst and posinst files, repackaging it and doing a ```dpkg -i <updated_pkg>.deb``` fixes this, but it's a very long and time-consuming process. 

Hence, since this PR is already raised, wanted to check if it is going to be merged as it would really be a big help

dmurphy18 commented:
@baby-gnu Still going through the PR, but wondering about relevance to Salt branches 3006.x and 3007.x, since it will be a while before the master branch gets turned into 3008, whether this PR should be applied to the 3006.x branch where the fixes will be forward merged, but leave that to your judgement. Asking since the referenced issue mentions 3006 and 3007 versions.

baby-gnu commented:
> @baby-gnu Still going through the PR, but wondering about relevance to Salt branches 3006.x and 3007.x, since it will be a while before the master branch gets turned into 3008, whether this PR should be applied to the 3006.x branch where the fixes will be forward merged, but leave that to your judgement. Asking since the referenced issue mentions 3006 and 3007 versions.

Ho, I did not understand that I should rebase the branch on `3006.x` instead of master, should I edit the PR to modify the base branch to `3006.x` too ?

dmurphy18 commented:
@baby-gnu Just thinking that it fi you want it to get out as a fix, rebasing to 3006.x will get it in a release quicker.
Discussed this in standup today and about to edit the pr template to suggest such, that is briefly, features go against master, bug fixes against 3006 or 3007 depending on the problem.

akshar-rafay commented:
@dmurphy18 @baby-gnu Sorry, just checking in again, any further update on when this is coming? And would it be to 3006.x or to master

baby-gnu commented:
> @dmurphy18 @baby-gnu Sorry, just checking in again, any further update on when this is coming? And would it be to 3006.x or to master

I'll rebase on `3006.x` soon.

dmurphy18 commented:
@baby-gnu Sorry don't understand your Q in https://github.com/saltstack/salt/pull/66688#issuecomment-2352568303
Master -> 3008 will be some time, not soon, hence recommendation for 3006.x.  Merge-forwards from branch 3006.x happen all the time.  Hope this answers your question.

baby-gnu commented:
I finally took time to rebase on 3006.x, I'll test the generated packages tomorrow.

baby-gnu commented:
I don't know why the `salt-cloud.postinst` [is failing](https://github.com/saltstack/salt/actions/runs/10990406711/job/30516721671?pr=66688), I can't reproduce locally

<details><summary>Error from <code>salt-cloud.postinst</code></summary>

```
Setting up salt-cloud (3006.9+151.gb522bac156) ...
debconf: unable to initialize frontend: Dialog
debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg: error processing package salt-cloud (--configure):
 installed salt-cloud package post-installation script subprocess returned error exit status 10
```
</details>

twangboy commented:
@dmurphy18 Would you have an idea what's going on here?

dmurphy18 commented:
@baby-gnu Set ```-x```` to get debug information output in the logs or screen, also make use of ```echo``` for the same reason. That is what I do to debug these.  About to head out for the day but will dig in , in the morning

baby-gnu commented:
> After greping my `/var/lib/dpkg/info/`, I choose to add a bunch of `|| true` 😉

So, after enclosing the fragile code in `if`, because `db_get` put its error message in `$RET`, my PR is ready.

Looking at the failing tests, I found that the tests [requires the user to be `salt`](https://github.com/saltstack/salt/blob/3006.x/tests/pytests/pkg/integration/test_salt_user.py#L91) but the actual behaviour is to [keep the current user on upgrade](https://github.com/saltstack/salt/blob/3006.x/pkg/debian/salt-master.preinst#L21) 🤔

baby-gnu commented:
> Yeah, The CI/CD testing infrastructure wants user `salt`. However in the wild, users can change that non-root user to whatever they want and when upgrading , the upgrade should not force the user to `salt` but leave it as the original user.

Thanks @dmurphy18.

> You have some test failures btw.

Yes, but I do not understand them yet

<details><summary>Error of <code>test_salt_user_master</code></summary>

```
    def test_salt_user_master(install_salt, salt_master):
        """
        Test the correct user is running the Salt Master
        """
        for count in range(0, 30):
            if salt_master.is_running():
                break
            else:
                time.sleep(2)

>       assert salt_master.is_running()
E       assert False
E        +  where False = <bound method Factory.is_running of SaltMaster(salt_pkg_install=SaltPkgInstall(pkg_system_service=True, root=None, run_root=None, ssm_bin=None, bin_dir=None, install_dir=PosixPath('/opt/saltstack/salt'), binary_paths={'salt': [PosixPath('/opt/saltstack/salt/salt')], 'api': [PosixPath('/opt/saltstack/salt/salt-api')], 'call': [PosixPath('/opt/saltstack/salt/salt-call')], 'cloud': [PosixPath('/opt/saltstack/salt/salt-cloud')], 'cp': [PosixPath('/opt/saltstack/salt/salt-cp')], 'key': [PosixPath('/opt/saltstack/salt/salt-key')], 'master': [PosixPath('/opt/saltstack/salt/salt-master')], 'minion': [PosixPath('/opt/saltstack/salt/salt-minion')], 'proxy': [PosixPath('/opt/saltstack/salt/salt-proxy')], 'run': [PosixPath('/opt/saltstack/salt/salt-run')], 'ssh': [PosixPath('/opt/saltstack/salt/salt-ssh')], 'syndic': [PosixPath('/opt/saltstack/salt/salt-syndic')], 'spm': [PosixPath('/opt/saltstack/salt/spm')], 'pip': [PosixPath('/opt/saltstack/salt/salt-pip')], 'python': [PosixPath('/opt/saltstack/salt/bin/python3')]}, config_path=PosixPath('/etc/salt'), conf_dir=PosixPath('/etc/salt'), upgrade=False, downgrade=False, classic=False, no_install=False, no_uninstall=False, distro_id='ubuntu', distro_codename='noble', distro_name='ubuntu', distro_version='24.04', prev_version=None, use_prev_version=False, artifact_version='3006.9+158.gea9cb10628', version='3006.9+158.gea9cb10628', pkg_mngr='apt-get', rm_pkg='purge', dbg_pkg='salt-dbg', salt_pkgs=['salt-api', 'salt-syndic', 'salt-ssh', 'salt-master', 'salt-cloud', 'salt-minion', 'salt-common', 'salt-dbg'], pkgs=['/tmp/testing/artifacts/pkg/salt-common_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-master_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-cloud_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-dbg_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-minion_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-api_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-syndic_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-ssh_3006.9+158.gea9cb10628_arm64.deb'], file_ext='deb', relenv=True), id='master-R4EmQu', config_file='/etc/salt/master', config_dir='/etc/salt', python_executable=None, cwd=PosixPath('/tmp/testing'), slow_stop=True, timeout=None, script_name='salt-master', base_script_args=[], check_ports=[], extra_cli_arguments_after_first_start_failure=[], display_name="SaltMaster(id='master-R4EmQu')")>()
E        +    where <bound method Factory.is_running of SaltMaster(salt_pkg_install=SaltPkgInstall(pkg_system_service=True, root=None, run_root=None, ssm_bin=None, bin_dir=None, install_dir=PosixPath('/opt/saltstack/salt'), binary_paths={'salt': [PosixPath('/opt/saltstack/salt/salt')], 'api': [PosixPath('/opt/saltstack/salt/salt-api')], 'call': [PosixPath('/opt/saltstack/salt/salt-call')], 'cloud': [PosixPath('/opt/saltstack/salt/salt-cloud')], 'cp': [PosixPath('/opt/saltstack/salt/salt-cp')], 'key': [PosixPath('/opt/saltstack/salt/salt-key')], 'master': [PosixPath('/opt/saltstack/salt/salt-master')], 'minion': [PosixPath('/opt/saltstack/salt/salt-minion')], 'proxy': [PosixPath('/opt/saltstack/salt/salt-proxy')], 'run': [PosixPath('/opt/saltstack/salt/salt-run')], 'ssh': [PosixPath('/opt/saltstack/salt/salt-ssh')], 'syndic': [PosixPath('/opt/saltstack/salt/salt-syndic')], 'spm': [PosixPath('/opt/saltstack/salt/spm')], 'pip': [PosixPath('/opt/saltstack/salt/salt-pip')], 'python': [PosixPath('/opt/saltstack/salt/bin/python3')]}, config_path=PosixPath('/etc/salt'), conf_dir=PosixPath('/etc/salt'), upgrade=False, downgrade=False, classic=False, no_install=False, no_uninstall=False, distro_id='ubuntu', distro_codename='noble', distro_name='ubuntu', distro_version='24.04', prev_version=None, use_prev_version=False, artifact_version='3006.9+158.gea9cb10628', version='3006.9+158.gea9cb10628', pkg_mngr='apt-get', rm_pkg='purge', dbg_pkg='salt-dbg', salt_pkgs=['salt-api', 'salt-syndic', 'salt-ssh', 'salt-master', 'salt-cloud', 'salt-minion', 'salt-common', 'salt-dbg'], pkgs=['/tmp/testing/artifacts/pkg/salt-common_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-master_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-cloud_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-dbg_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-minion_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-api_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-syndic_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-ssh_3006.9+158.gea9cb10628_arm64.deb'], file_ext='deb', relenv=True), id='master-R4EmQu', config_file='/etc/salt/master', config_dir='/etc/salt', python_executable=None, cwd=PosixPath('/tmp/testing'), slow_stop=True, timeout=None, script_name='salt-master', base_script_args=[], check_ports=[], extra_cli_arguments_after_first_start_failure=[], display_name="SaltMaster(id='master-R4EmQu')")> = SaltMaster(salt_pkg_install=SaltPkgInstall(pkg_system_service=True, root=None, run_root=None, ssm_bin=None, bin_dir=None, install_dir=PosixPath('/opt/saltstack/salt'), binary_paths={'salt': [PosixPath('/opt/saltstack/salt/salt')], 'api': [PosixPath('/opt/saltstack/salt/salt-api')], 'call': [PosixPath('/opt/saltstack/salt/salt-call')], 'cloud': [PosixPath('/opt/saltstack/salt/salt-cloud')], 'cp': [PosixPath('/opt/saltstack/salt/salt-cp')], 'key': [PosixPath('/opt/saltstack/salt/salt-key')], 'master': [PosixPath('/opt/saltstack/salt/salt-master')], 'minion': [PosixPath('/opt/saltstack/salt/salt-minion')], 'proxy': [PosixPath('/opt/saltstack/salt/salt-proxy')], 'run': [PosixPath('/opt/saltstack/salt/salt-run')], 'ssh': [PosixPath('/opt/saltstack/salt/salt-ssh')], 'syndic': [PosixPath('/opt/saltstack/salt/salt-syndic')], 'spm': [PosixPath('/opt/saltstack/salt/spm')], 'pip': [PosixPath('/opt/saltstack/salt/salt-pip')], 'python': [PosixPath('/opt/saltstack/salt/bin/python3')]}, config_path=PosixPath('/etc/salt'), conf_dir=PosixPath('/etc/salt'), upgrade=False, downgrade=False, classic=False, no_install=False, no_uninstall=False, distro_id='ubuntu', distro_codename='noble', distro_name='ubuntu', distro_version='24.04', prev_version=None, use_prev_version=False, artifact_version='3006.9+158.gea9cb10628', version='3006.9+158.gea9cb10628', pkg_mngr='apt-get', rm_pkg='purge', dbg_pkg='salt-dbg', salt_pkgs=['salt-api', 'salt-syndic', 'salt-ssh', 'salt-master', 'salt-cloud', 'salt-minion', 'salt-common', 'salt-dbg'], pkgs=['/tmp/testing/artifacts/pkg/salt-common_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-master_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-cloud_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-dbg_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-minion_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-api_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-syndic_3006.9+158.gea9cb10628_arm64.deb', '/tmp/testing/artifacts/pkg/salt-ssh_3006.9+158.gea9cb10628_arm64.deb'], file_ext='deb', relenv=True), id='master-R4EmQu', config_file='/etc/salt/master', config_dir='/etc/salt', python_executable=None, cwd=PosixPath('/tmp/testing'), slow_stop=True, timeout=None, script_name='salt-master', base_script_args=[], check_ports=[], extra_cli_arguments_after_first_start_failure=[], display_name="SaltMaster(id='master-R4EmQu')").is_running
```
</details>

dmurphy18 commented:
@baby-gnu I see those testing infrastructure failures occasionally too , but have not been able to find out what is going on, since the salt-master or salt-minion is supposed to be running  when the test is entered and that is basically a sanity check to catch when not and try to figure what is going on when it happens.

That is why the count code was entered to see if this was some time delay issue, which helped initially, but I am wondering what is going on behind the scenes on GitHub, since seen the same test take 25 minutes and then take 45 minutes, reason had to up the time slice for tests from 120 to 180 for MacOS. Are they stealing cycles behind the scene but still using a real-time clock, don't know.

baby-gnu commented:
For the failing tests, I'm scratching my head:

- tests define a [list of directories owned by `salt`](https://github.com/saltstack/salt/blob/3006.x/tests/pytests/pkg/integration/test_salt_user.py#L56)
- `salt-master.preinst` was [fix old package (3006.0/3006.1) perms that used the `salt` user too widely](https://github.com/saltstack/salt/commit/76df410a367a1406a8a29f6573e15c3a03769a96)

I think I'll stick to the test file being the canonical choice and fix the `.preinst`.

baby-gnu commented:
I found my issue, it was because the salt-master configuration must be define before `salt-cloud` call to `db_get salt-master/user`.

I'll rebase the branch to sign commits and remove the `-x` debug flag when ready to be merged.

dmurphy18 commented:
@baby-gnu FYI salt-api, salt-syndic and salt-cloud all require salt-master before installing, thought this was well-known and stated so in the docs

baby-gnu commented:
> @baby-gnu FYI salt-api, salt-syndic and salt-cloud all require salt-master before installing, thought this was well-known and stated so in the docs

Yes, I verified, now the packaging will always order correctly the packaging scripts.

@dwoz should I still rebase the branch from time to time to update it?

dmurphy18 reviewed (dismissed):
Understand need for 'set -e -x' when developing, but need to remove '-x' before this PR can be merged.  Cannot have debug information going to the screen etc.

And would rebase against current branch after updating/adjusting PR

baby-gnu commented:
> Understand need for 'set -e -x' when developing, but need to remove '-x' before this PR can be merged. Cannot have debug information going to the screen etc.
> 
> And would rebase against current branch after updating/adjusting PR

Thanks @dmurphy18, I remove the commit which added the `-x` and rebased the branch.

dwoz commented:
> > @baby-gnu FYI salt-api, salt-syndic and salt-cloud all require salt-master before installing, thought this was well-known and stated so in the docs
> 
> Yes, I verified, now the packaging will always order correctly the packaging scripts.
> 
> @dwoz should I still rebase the branch from time to time to update it?

Yes, please keep it up to date if possible.

baby-gnu commented:
The failing tests don't seem to be related to my PR, right?

baby-gnu commented:
The failing tests seems not related to this PR.

baby-gnu commented:
Do you think this could reach the [next point release](https://saltproject.io/blog/2025-01-16-open-hour/#upcoming-salt-3006-lts--3007-sts-point-releases)?

baby-gnu commented:
I found new corner cases where some `chown` in `preinst` where failing when a service was never started.

baby-gnu committed:
Make deb scripts more compliant with Debian policy

baby-gnu commented:
Thanks @twangboy for the approval, I keep maintaining the branch up-to-date hoping for being in the next point release :sweat_smile:

baby-gnu commented:
All checks pass :tada:

dmurphy18 commented:
Moving this to 3008 milestone, for now, since don't want to destabilize what is being worked on at the moment, will review and possible move back to 3006.x branch milestone, in a week or two's time.

baby-gnu commented:
@dmurphy18 Ok, I'm waiting for your GO to rebase the branch, I let it outdated for now to avoid taking runner time for nothing.

baby-gnu commented:
Does it mean that it will not be for the next 3006?

dmurphy18 reviewed (dismissed):
Researching db_installdeb, need to update for latest branch changes to 3006.x, added test:full to ensure tested on all Debian and Ubuntu platforms
Assigned to both of us, so I can track issues.

barneysowood commented on pkg/debian/salt-master.preinst:
```diff
@@ -1,47 +1,67 @@
-    [ -z "$SALT_GROUP" ] && SALT_GROUP=salt
     PY_VER=$(/opt/saltstack/salt/bin/python3 -c "import sys; sys.stdout.write('{}.{}'.format(*sys.version_info)); sys.stdout.flush();")
 
     # Reset permissions to fix previous installs
```
I'd suggest removing this find/chmod entirely - the original purpose of this was broken in #66218 - the purpose was to reset these directories to being owned by root:root so that only the minimum set of dirs that needed to be owned by the salt user would be set in the postinst

This is essentially a no-op currently, it finds files/dirs owned by salt or group salt and chowns them to salt:salt.

barneysowood reviewed (dismissed):
# Removing checks for current service state

You're removing the checks that @dmurphy18 added in #66218 to get the current state (active/inactive and enabled/disabled) of the systemd services (salt-master, salt-api, salt-minion, salt-syndic).

I think there's a good argument that what you're doing is actually correct. The way that an end user should handle not wanting the service running is to mask it and the package shouldn't work out current state and try to preserve that.

However.. those changes *were* introduced in #66218 and if we change that behaviour now, that's going to mean people upgrading to 3006.11 (if that happens) will potentially get a surprise - services they had deactivated will be started.

I suspect @dmurphy18 would have argued we need to keep the existing behaviour because we can't break it now. I also suspect that may have been driven by his interactions with vmware customers - I can't speak to that, but perhaps @twangboy or @ScriptAutomate can chime in on that?

My suggestion would be that we do remove this and rely on users masking the service, *but* we don't do that in 3006.x. It should wait until 3008 and be clearly documented as a breaking change. We should probably also make the RPM packages consistent.

# Tests

This needs some tests to ensure we don't get a regression on this in the future. 

I did wonder why this wasn't failing tests - looks like the relevant tests have been disabled and haven't been re-enabled when they should have been - I've opened #67968 to resolve that.

barneysowood commented on pkg/debian/salt-master.preinst:
```diff
@@ -1,47 +1,67 @@
         /opt/saltstack/salt/lib/python${PY_VER}/site-packages/salt/cloud/deploy\* \) -a \( -user ${SALT_USER} \
-        -o -group ${SALT_GROUP} \) -exec chown ${SALT_USER}:${SALT_GROUP} \{\} \;
-
+        -o -group ${SALT_GROUP} \) -exec chown ${SALT_USER}:${SALT_GROUP} \{\} \; \
```
I'd suggest removing this find/chmod entirely - the original purpose of this was broken in #66218 - the purpose was to reset these directories to being owned by root:root so that only the minimum set of dirs that needed to be owned by the salt user would be set in the postinst

This is essentially a no-op currently, it finds files/dirs owned by salt or group salt and chowns them to salt:salt.

baby-gnu replied:
ok, I'll do that.

dafyddj commented:
Agree that any behaviour change should be made in a major release.

dwoz reviewed (approved):
@baby-gnu Can you please rebase this one?

dwoz reviewed (dismissed):
I think we absolutely need to have this addressed. I also think now that the package tests under `tests/pytest/pkg` are working. There should be something added to validate the changes here.

@baby-gnu Are there functionality changes that should go into anothar branch or not?

baby-gnu commented:
@baby-gnu to me, there is no breaking changes:

- enabled and running services are restarted
- disabled packages stay disabled
- masked packages stay masked

barneysowood commented:
> @baby-gnu to me, there is no breaking changes:
> 
> * enabled and running services are restarted
> * disabled packages stay disabled
> * masked packages stay masked

Masked services stay masked, but disabled or inactive services are running and no longer disabled after the upgrade. To be sure I tested it with a package artifact from this PR.

That's a change in behaviour to 3006.10...

baby-gnu committed:
No need to reset owner on master install

baby-gnu commented:
> > @baby-gnu to me, there is no breaking changes:
> > 
> > * enabled and running services are restarted
> > * disabled packages stay disabled
> > * masked packages stay masked
> 
> Masked services stay masked, but disabled or inactive services are running and no longer disabled after the upgrade. To be sure I tested it with a package artifact from this PR.
> 
> That's a change in behaviour to 3006.10...

I found the issue: the previous packages did not set the `state` file of `deb-systemd-helper` and the corresponding service is considered enabled.

I'll add a preinst test to create that file manually if required.

baby-gnu committed:
Work around deb package service status detection

Service installed without debhelper has no state file and the status
detection mechanism defaults to enabled.

Create the debhelper state file during package preinst if required.
